File: backend/src/handlers/providers.rs
use axum::{
    Json,
    extract::{Path, Query, State},
};
use serde::Deserialize;
use uuid::Uuid;

use crate::{
    db::{
        provider_endpoints::ProviderEndpoint, provider_keys::ProviderKey, providers::Provider,
        types::ApiType,
    },
    error::{AppError, AppResult},
    services::providers,
    state::AppState,
};

#[derive(Debug, Deserialize)]
pub struct ListProvidersQuery {
    #[serde(default = "default_page")]
    pub page: i64,
    #[serde(default = "default_page_size")]
    pub page_size: i64,
}

fn default_page() -> i64 {
    1
}

fn default_page_size() -> i64 {
    20
}

pub async fn list_providers(
    State(state): State<AppState>,
    Query(query): Query<ListProvidersQuery>,
) -> AppResult<Json<Vec<Provider>>> {
    let providers = providers::list_providers(&state.pool, query.page, query.page_size).await?;
    Ok(Json(providers))
}

#[derive(Debug, Deserialize)]
pub struct CreateProviderRequest {
    pub name: String,
    pub description: Option<String>,
    pub brief: Option<String>,
}

pub async fn create_provider(
    State(state): State<AppState>,
    Json(payload): Json<CreateProviderRequest>,
) -> AppResult<Json<Provider>> {
    let provider = providers::create_provider(
        &state.pool,
        payload.name,
        payload.description,
        payload.brief,
    )
    .await?;
    Ok(Json(provider))
}

pub async fn get_provider(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
) -> AppResult<Json<Provider>> {
    let provider = providers::get_provider(&state.pool, id)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(provider))
}

#[derive(Debug, Deserialize)]
pub struct UpdateProviderRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub brief: Option<String>,
    pub enabled: Option<bool>,
}

pub async fn update_provider(
    State(state): State<AppState>,
    Path(id): Path<Uuid>,
    Json(payload): Json<UpdateProviderRequest>,
) -> AppResult<Json<Provider>> {
    let provider = providers::update_provider(
        &state.pool,
        id,
        payload.name,
        payload.description,
        payload.brief,
        payload.enabled,
    )
    .await?
    .ok_or(AppError::NotFound)?;
    Ok(Json(provider))
}

pub async fn delete_provider(State(state): State<AppState>, Path(id): Path<Uuid>) -> AppResult<()> {
    if providers::delete_provider(&state.pool, id).await? {
        Ok(())
    } else {
        Err(AppError::NotFound)
    }
}

// Endpoints

pub async fn list_endpoints(
    State(state): State<AppState>,
    Path(provider_id): Path<Uuid>,
) -> AppResult<Json<Vec<ProviderEndpoint>>> {
    let endpoints = providers::list_endpoints_by_provider(&state.pool, provider_id).await?;
    Ok(Json(endpoints))
}

#[derive(Debug, Deserialize)]
pub struct CreateEndpointRequest {
    pub api_type: ApiType,
    pub url: String,
}

pub async fn create_endpoint(
    State(state): State<AppState>,
    Path(provider_id): Path<Uuid>,
    Json(payload): Json<CreateEndpointRequest>,
) -> AppResult<Json<ProviderEndpoint>> {
    let endpoint =
        providers::create_endpoint(&state.pool, provider_id, payload.api_type, payload.url).await?;
    Ok(Json(endpoint))
}

#[derive(Debug, Deserialize)]
pub struct UpdateEndpointRequest {
    pub url: Option<String>,
    pub enabled: Option<bool>,
}

pub async fn update_endpoint(
    State(state): State<AppState>,
    Path((_provider_id, id)): Path<(Uuid, Uuid)>,
    Json(payload): Json<UpdateEndpointRequest>,
) -> AppResult<Json<ProviderEndpoint>> {
    let endpoint = providers::update_endpoint(&state.pool, id, payload.url, payload.enabled)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(endpoint))
}

pub async fn delete_endpoint(
    State(state): State<AppState>,
    Path((_provider_id, id)): Path<(Uuid, Uuid)>,
) -> AppResult<()> {
    if providers::delete_endpoint(&state.pool, id).await? {
        Ok(())
    } else {
        Err(AppError::NotFound)
    }
}

// Keys

pub async fn list_keys(
    State(state): State<AppState>,
    Path(provider_id): Path<Uuid>,
) -> AppResult<Json<Vec<ProviderKey>>> {
    let keys = providers::list_keys_by_provider(&state.pool, provider_id).await?;
    Ok(Json(keys))
}

#[derive(Debug, Deserialize)]
pub struct CreateKeyRequest {
    pub name: Option<String>,
    pub key: String,
}

pub async fn create_key(
    State(state): State<AppState>,
    Path(provider_id): Path<Uuid>,
    Json(payload): Json<CreateKeyRequest>,
) -> AppResult<Json<ProviderKey>> {
    let key = providers::create_key(&state.pool, provider_id, payload.name, payload.key).await?;
    Ok(Json(key))
}

#[derive(Debug, Deserialize)]
pub struct UpdateKeyRequest {
    pub name: Option<String>,
    pub enabled: Option<bool>,
}

pub async fn update_key(
    State(state): State<AppState>,
    Path((_provider_id, id)): Path<(Uuid, Uuid)>,
    Json(payload): Json<UpdateKeyRequest>,
) -> AppResult<Json<ProviderKey>> {
    let key = providers::update_key(&state.pool, id, payload.name, payload.enabled)
        .await?
        .ok_or(AppError::NotFound)?;
    Ok(Json(key))
}

pub async fn delete_key(
    State(state): State<AppState>,
    Path((_provider_id, id)): Path<(Uuid, Uuid)>,
) -> AppResult<()> {
    if providers::delete_key(&state.pool, id).await? {
        Ok(())
    } else {
        Err(AppError::NotFound)
    }
}

// Models

pub async fn list_provider_models(
    State(state): State<AppState>,
    Path(provider_id): Path<Uuid>,
) -> AppResult<Json<Vec<crate::services::openai::Model>>> {
    let models = state.openai.list_models(provider_id).await?;
    Ok(Json(models))
}

File: frontend/src/api/providers.ts
import { requestJson } from "./client";

export type ApiType = 'openai_chat_completions' | 'openai_responses' | 'openai_models' | 'anthropic_messages';

export interface Provider {
  id: string;
  name: string;
  description: string | null;
  brief: string | null;
  enabled: boolean;
  created_at: string;
}

export interface ProviderEndpoint {
  id: string;
  provider_id: string;
  api_type: ApiType;
  url: string;
  enabled: boolean;
  created_at: string;
}

export interface ProviderKey {
  id: string;
  provider_id: string;
  name: string | null;
  key: string;
  usage_count: number;
  enabled: boolean;
  created_at: string;
}

export interface ListProvidersParams {
  page?: number;
  page_size?: number;
}

export interface CreateProviderRequest {
  name: string;
  description?: string;
  brief?: string;
}

export interface UpdateProviderRequest {
  name?: string;
  description?: string;
  brief?: string;
  enabled?: boolean;
}

export interface CreateEndpointRequest {
  api_type: ApiType;
  url: string;
}

export interface UpdateEndpointRequest {
  url?: string;
  enabled?: boolean;
}

export interface CreateKeyRequest {
  name?: string;
  key: string;
}

export interface UpdateKeyRequest {
  name?: string;
  enabled?: boolean;
}

// Provider API
export async function listProviders(params: ListProvidersParams = {}): Promise<Provider[]> {
  const query = new URLSearchParams();
  if (params.page !== undefined) query.set("page", params.page.toString());
  if (params.page_size !== undefined) query.set("page_size", params.page_size.toString());

  const queryString = query.toString();
  return requestJson<Provider[]>(`/providers${queryString ? `?${queryString}` : ""}`);
}

export async function createProvider(payload: CreateProviderRequest): Promise<Provider> {
  return requestJson<Provider>("/providers", {
    method: "POST",
    body: payload,
  });
}

export async function getProvider(id: string): Promise<Provider> {
  return requestJson<Provider>(`/providers/${id}`);
}

export async function updateProvider(id: string, payload: UpdateProviderRequest): Promise<Provider> {
  return requestJson<Provider>(`/providers/${id}`, {
    method: "PUT",
    body: payload,
  });
}

export async function deleteProvider(id: string): Promise<void> {
  await requestJson<void>(`/providers/${id}`, {
    method: "DELETE",
  });
}

// Endpoint API
export async function listEndpoints(providerId: string): Promise<ProviderEndpoint[]> {
  return requestJson<ProviderEndpoint[]>(`/providers/${providerId}/endpoints`);
}

export async function createEndpoint(providerId: string, payload: CreateEndpointRequest): Promise<ProviderEndpoint> {
  return requestJson<ProviderEndpoint>(`/providers/${providerId}/endpoints`, {
    method: "POST",
    body: payload,
  });
}

export async function updateEndpoint(providerId: string, endpointId: string, payload: UpdateEndpointRequest): Promise<ProviderEndpoint> {
  return requestJson<ProviderEndpoint>(`/providers/${providerId}/endpoints/${endpointId}`, {
    method: "PUT",
    body: payload,
  });
}

export async function deleteEndpoint(providerId: string, endpointId: string): Promise<void> {
  await requestJson<void>(`/providers/${providerId}/endpoints/${endpointId}`, {
    method: "DELETE",
  });
}

// Key API
export async function listKeys(providerId: string): Promise<ProviderKey[]> {
  return requestJson<ProviderKey[]>(`/providers/${providerId}/keys`);
}

export async function createKey(providerId: string, payload: CreateKeyRequest): Promise<ProviderKey> {
  return requestJson<ProviderKey>(`/providers/${providerId}/keys`, {
    method: "POST",
    body: payload,
  });
}

export async function updateKey(providerId: string, keyId: string, payload: UpdateKeyRequest): Promise<ProviderKey> {
  return requestJson<ProviderKey>(`/providers/${providerId}/keys/${keyId}`, {
    method: "PUT",
    body: payload,
  });
}

export async function deleteKey(providerId: string, keyId: string): Promise<void> {
  await requestJson<void>(`/providers/${providerId}/keys/${keyId}`, {
    method: "DELETE",
  });
}